doc: update content on dockerTools.buildImage and follow doc conventions (#278956)
NixOS/nixpkgs: doc/build-helpers/images/dockertools.section.md

    WorkingDir = "/data";
    Volumes = { "/data" = { }; };
  };
}
```

> 

Reviewer: Would be good to show how to load it into docker here.

  Otherwise, the resulting image with contain the single layer created by `buildImage`.

    _Default value:_ `null`.

`fromImageName` _optional_

: Used to specify the image within the repository tarball in case it contains multiple images.
  A value of `null` means that `buildImage` will use the first image available in the repository.

  :::{.note}
  This must be used with `fromImageTag`. Using only `fromImageName` without `fromImageTag` will make `buildImage` use the first image available in the repository.
  :::

    _Default value:_ `null`.

`fromImageTag` _optional_

: Used to specify the image within the repository tarball in case it contains multiple images.
  A value of `null` means that `buildImage` will use the first image available in the repository.

  :::{.note}
  This must be used with `fromImageName`. Using only `fromImageTag` without `fromImageName` will make `buildImage` use the first image available in the repository
  :::

    _Default value:_ `null`.

`copyToRoot` _optional_

: Files to add to the generated image.

>   This can be either a path or a list of paths.

Author: I used "path" to mean anything that coerces to a path, but I'll make this explicit.

  This can be seen as an equivalent of `ADD contents/ /` in a `Dockerfile`.

  If a derivation is used in this attribute, its output path will be used instead.

>   Because of this, it's a common usage to see `pkgs.buildEnv` used in this attribute.

Reviewer 2: > thoughts

Can't block on this comment, because it balloons the scope of the PR insanely. Feel free to ignore and just clarify the text without referring to anything.

>   Because of this, it's a common usage to see `pkgs.buildEnv` used in this attribute.

Author: > I think we should always link to a page (e.g. in the Nix manual or nix.dev) that clarifies and disambiguates terms, as well as serve as a writing guide

I agree with this. I'm already working on the next section of `dockerTools` and encountered a similar issue when talking about Nix store objects and Nix store paths.

Given that the content related to contributing was moved out of the Nix manual, I think nix.dev is a more appropriate place to put this guide. I'll treat it as a separate piece of work not linked to this one to keep the scope of this PR focused.

>   Because of this, it's a common usage to see `pkgs.buildEnv` used in this attribute.

Reviewer 3: It would make sense to introduce these terms in some sort of orientation article on nix.dev, but really this is a matter of definitions, so it should find a place in the Nixpkgs manual (which prominently doesn't have a glossary yet).

    _Default value:_ `null`.

`keepContentsDirlinks` _optional_

: When adding files to the generated image (as specified by `copyToRoot`), this attribute controls whether to preserve symlinks to directories.
  If `false`, the symlinks will be transformed into directories.
  This behaves the same as `rsync -k` when `keepContentsDirlinks` is `false`, and the same as `rsync -K` when `keepContentsDirlinks` is `true`.

    _Default value:_ `false`.

`runAsRoot` _optional_

: A bash script that will run as root inside a VM that contains the existing layers of the base image and the new generated layer (including the files from `copyToRoot`).
  This can be seen as an equivalent of `RUN ...` in a `Dockerfile`.
  A value of `null` means that this step in the image generation process will be skipped.

  :::{.caution}

>   Using this attribute requires the `kvm` device to be available.

Reviewer 3: ```suggestion
  Using this attribute requires the `kvm` device to be available, see [`system-features`](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-system-features)
```

>   Using this attribute requires the `kvm` device to be available.

Reviewer: The highlight in this case actually kind of makes sense, I didn't mind keeping it :)

>   Using this attribute requires the `kvm` device to be available.

Author: Decided to use the link without highlight. After opening the one with highlight again, it felt a bit weird because only 2 words in the whole page were highlighted.

  It is simply run as part of building the derivation that outputs the layer created by `buildImage`.

    _Default value:_ `""`.

`config` _optional_

: Used to specify the configuration of the containers that will be started off the generated image.
  Must be an attribute set, with each attribute as listed in the [Docker Image Specification v1.3.0](https://github.com/moby/moby/blob/46f7ab808b9504d735d600e259ca0723f76fb164/image/spec/spec.md#image-json-field-descriptions).

    _Default value:_ `null`.

`architecture` _optional_

: Used to specify the image architecture.
  This is useful for multi-architecture builds that don't need cross compiling.
  If specified, its value should follow the [OCI Image Configuration Specification](https://github.com/opencontainers/image-spec/blob/main/config.md#properties), which should still be compatible with Docker.

    _Default value:_ the same value from `pkgs.go.GOARCH`.

`diskSize` _optional_

: Controls the disk size (in megabytes) of the VM used to run the script specified in `runAsRoot`.
  This attribute is ignored if `runAsRoot` is `null`.

    _Default value:_ 1024.

`buildVMMemorySize` _optional_

> : Controls the amount of memory (in megabytes) provisioend for the VM used to run the script specified in `runAsRoot`.

Reviewer: ```suggestion
: Controls the amount of memory (in megabytes) provisioned for the VM used to run the script specified in `runAsRoot`.
```

  This attribute is ignored if `runAsRoot` is `null`.

    _Default value:_ 512.

`created` _optional_

: Specifies the time of creation of the generated image.

>   This should be either an ISO-8601-formatted date and time or `"now"`, in which case `buildImage` will use the current date.

Reviewer: If you can find a good link to the ISO8601 spec, would be good to add here

Reviewer 3: ```suggestion
  This should be either a date and time formatted according to [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601)  or `"now"`, in which case `buildImage` will use the current date.
```

  :::{.caution}
  Using `"now"` means that the generated image will not be reproducible anymore (because the date will always change whenever it's built).
  :::

    _Default value:_ `"1970-01-01T00:00:01Z"`.

`uid` _optional_

: The uid of the user that will own the files packed in the new layer built by `buildImage`.

    _Default value:_ 0.

`gid` _optional_

: The gid of the group that will own the files packed in the new layer built by `buildImage`.

    _Default value:_ 0.

`contents` **DEPRECATED**

: This attribute is deprecated, and users are encouraged to use `copyToRoot` instead.

> `buildImage` defines a few [`passthru`](#var-stdenv-passthru) attributes:

Reviewer: It would be good to have a cleaner separation from inputs vs "outputs", maybe a nested heading for each would work.
